### PR TITLE
Feat/setup locker

### LINKED ISSUE #10: Implement the "Setup locker" user-journey
## Description

Implement the "Setup locker" screen.

## Acceptance Criteria

- [x] Implement the interface as specified in #9

### PR DESCRIPTION
**Pull request by dominguesgm**
Implements the Setup locker user journey.

Closes https://github.com/ipfs-shipyard/pm-idm/issues/10

### CODE REVIEW

**Comment by satazor** [comment]
Is it supposed to have a red line right away in the confirm password field?
Screen Shot 2019-04-29 at 13 14 34

**Comment by satazor** [comment]
Nvm, I was running an older version of the UIKit

File: src/shared/components/setup-locker-demo/SetupLocker.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React, { Component } from 'react';
```
Comment (satazor) [inline src/shared/components/setup-locker-demo/SetupLocker.js]: Can you delete all the demo files?

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    render() {
+        const { timeoutValue, feedback } = this.state;
+
+        return (
+            <div className={ styles.contentWrapper }>
+                <h2 className={ styles.title }>Setup expiration time.</h2>
+                <p>You may tweak the amount of time you are ablo to use
+                 the app before it locks again. You can go have a coffee without any worries.</p>
+                <div className={ styles.idlePicker }>
+                    <IdlePicker defaultValue={ timeoutValue } handlePickerChange={ this.onPickerChange } />
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: `handlePickerChange` was changed to `onChange` in IdlePicker.

Also the callback should be named `handlePickerChange` and no `onPickerChange`. For reference, callback props always start with `onXXX` and the handlers with `handleXXXX`.

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                </div>
+                <div className={ styles.continueButton }>
+                    <Button onClick={ this.handleContinue } onAnimationEnd={ this.handleButtonAnimationEnd }
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: Reminder: change `onAnimationEnd` to `onFeedbackAnimationEnd`.

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    onPickerChange = (value) => {
+        this.setState({ timeoutValue: value });
+    };
+
+    handleContinue = async () => {
+        const { timeoutValue } = this.state;
+        const { setMaxTime } = this.props;
+
+        try {
+            await setMaxTime(timeoutValue * 60 * 1000);
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: We should be careful when doing things like this: what happens if the component is unmounted while the promise is still resolving? If that happens, the `setState` will be called on a unmounted component.

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        } else {
+            this.setState({ feedback: 'none' });
+        }
+    };
+}
+
+SetTimeout.propTypes = {
+    onNextStep: PropTypes.func.isRequired,
+    feedback: PropTypes.oneOf(['none', 'loading', 'success', 'error']),
+    setMaxTime: PropTypes.func.isRequired,
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: Can you rename to `onSetMaxTime`? The same applies to other similar cases in SetPassphrase.

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        if (isSuccess) {
+            onNextStep();
+        } else {
+            this.setState({ feedback: 'none' });
+        }
+    };
+}
+
+SetTimeout.propTypes = {
+    onNextStep: PropTypes.func.isRequired,
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: Can you move this prop to last position?

File: src/shared/utils/utils.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+export const injectPropsIntoStepArray = (steps, stepId, props) => {
```
Comment (satazor) [inline src/shared/utils/utils.js]: What's this?

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/shared/components/lock-screen/LockScreen.js [hunk 8]
```diff
@@ -42,7 +42,9 @@ class LockScreen extends Component {
 
         input.focus();
         input.addEventListener('keydown', this.handleKeyboardInput);
-        input.addEventListener('blur', () => input.focus());
```
Comment (satazor) [inline src/shared/components/lock-screen/LockScreen.js]: Why was this changed?

File: src/shared/components/lock-screen/LockScreen.js [hunk 9]
```diff
@@ -128,6 +132,9 @@ class LockScreen extends Component {
                 this.handleSubmission();
             }
             this.passwordInputRef.current.focus();
+            setTimeout(() => {
```
Comment (satazor) [inline src/shared/components/lock-screen/LockScreen.js]: Why do we need this? Isn't `input.addEventListener('blur', () => input.focus());` enough?

File: src/shared/components/setup-locker/SetupLocker.js [hunk 10]
```diff
@@ -1,52 +1,63 @@
 import React, { Component } from 'react';
+import { FlowModal, FlowModalStep, setAppElement } from '@nomios/web-uikit';
 import PropTypes from 'prop-types';
 
-import SetupMasterLock from './steps/SetupMasterLock';
-import SetupIdleTimer from './steps/SetupIdleTimer';
+import SetPassphrase from './setup-locker-steps/set-passphrase/SetPassphrase';
+import SetTimeout from './setup-locker-steps/set-timeout/SetTimeout';
+
+setAppElement('#root');
```
Comment (satazor) [inline src/shared/components/setup-locker/SetupLocker.js]: Move this to `src/index.js` right above `ReactDOM.render(<App />, document.getElementById('root'));`

File: src/shared/components/setup-locker/SetupLocker.js [hunk 11]
```diff
@@ -1,52 +1,63 @@
...
     render() {
+        const { stepId } = this.state;
+        const { isOpen } = this.props;
+
         return (
-            <div>
-                <h1>Setup Locker</h1>
-                { this.renderStep() }
-            </div>
+            <FlowModal open={ isOpen } variant="simple" step={ stepId } showClose={ false }>
```
Comment (satazor) [inline src/shared/components/setup-locker/SetupLocker.js]: Other Modal props should be supported and spread here. For instance, we will need to attach `onExited` to open the create/import flow afterwards.

File: src/shared/components/setup-locker/SetupLocker.js [hunk 12]
```diff
@@ -1,52 +1,63 @@
...
+
     state = {
-        step: 1,
+        stepId: 'passphrase',
+        data: {},
     };
 
     render() {
+        const { stepId } = this.state;
+        const { isOpen } = this.props;
```
Comment (satazor) [inline src/shared/components/setup-locker/SetupLocker.js]: Change to `open` for consistency with modals. See my comment below, because we might remove this if we use spread.

File: src/shared/components/setup-locker/SetupLocker.js [hunk 13]
```diff
@@ -1,52 +1,63 @@
...
-        if (step === 2) {
-            return onComplete();
-        }
+    handleProceedFromPassphrase = () => {
+        this.setState({ stepId: 'timeout' });
+    };
 
-        this.setState({ step: step + 1 });
+    handleProceedFromTimeout = () => {
+        setTimeout(this.props.onComplete, 300);
```
Comment (satazor) [inline src/shared/components/setup-locker/SetupLocker.js]: Why do we need the timeout?

File: src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js [hunk 14]
```diff
@@ -0,0 +1,253 @@
+import React, { Component } from 'react';
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: There's a lot of things going on here. Can we discuss this personally or in a call?

File: src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        const { analyzePasswordQuality } = this.props;
+        const passwordCopy = this.password;
+        const confirmationCopy = this.confirmation;
+
+        analyzePasswordQuality(passwordCopy)
+        .then((result) => this.handleSuccess(passwordCopy, confirmationCopy, result))
+        .catch((error) => this.handleError(passwordCopy, confirmationCopy, error));
+    };
+
+    validatePassword = () => {
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: Please use normal functions whenever possible. Make this change everywhere.

**Review by satazor — CHANGES_REQUESTED** [review]

File: src/shared/components/setup-locker-demo/SetupLocker.js (hunk 1 above)
Comment (dominguesgm) [reply src/shared/components/setup-locker-demo/SetupLocker.js]: Yes, I forgot to do it. At this point we can get rid of the demo files we needed to have earlier for testing purposes

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js (hunk 4 above)
Comment (dominguesgm) [reply src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: What would be the best approach to prevent that scenario?

File: src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js (hunk 5 above)
Comment (dominguesgm) [reply src/shared/components/setup-locker/setup-locker-steps/set-timeout/SetTimeout.js]: I can but the linter will complain about calling the callback in SetupLocker `setMaxTime`. Should I change the callback to `handleMaxTime`?

File: src/shared/utils/utils.js (hunk 7 above)
Comment (dominguesgm) [reply src/shared/utils/utils.js]: This was something needed for the previous approach of defining an object with all the steps information. As it is no longer needed, I'll be deleting it.

File: src/shared/components/lock-screen/LockScreen.js (hunk 8 above)
Comment (dominguesgm) [reply src/shared/components/lock-screen/LockScreen.js]: For testing purposes, I had to log this arrow function. That is the reason for that block being there. I'll change it back to the original form

File: src/shared/components/lock-screen/LockScreen.js (hunk 9 above)
Comment (dominguesgm) [reply src/shared/components/lock-screen/LockScreen.js]: Chrome was having some issues and launching a blur event at the end of the entire blur/focus flow when the user pressed `enter`, so I had to force a delayed focus to fix that.

**Comment by satazor** [comment]
The following error/warning is happening as soon as I hit "Complete":

> Once the modal is open, the first step must be rendered

**Comment by dominguesgm** [comment]
> The following error/warning is happening as soon as I hit "Complete":
> 
> > Once the modal is open, the first step must be rendered

File: src/shared/components/setup-locker/SetupLocker.js [hunk 16]
```diff
@@ -1,52 +1,63 @@
...
-                <h1>Setup Locker</h1>
-                { this.renderStep() }
-            </div>
+            <FlowModal open={ isOpen } variant="simple" step={ stepId } showClose={ false }>
+                <FlowModalStep id="passphrase">
+                    <SetPassphrase onNextStep={ this.handleProceedFromPassphrase }
+                        analyzePasswordQuality={ this.analyzePasswordQuality } enablePassword={ this.enablePassword } />
+                </FlowModalStep>
+                <FlowModalStep id="timeout">
+                    <SetTimeout onNextStep={ this.handleProceedFromTimeout } setMaxTime={ this.setMaxTime } />
```
Comment (satazor) [inline src/shared/components/setup-locker/SetupLocker.js]: Change SetTimeout to SetIdleTimer

File: src/shared/components/setup-locker/SetupLocker.js (hunk 13 above)
Comment (satazor) [reply src/shared/components/setup-locker/SetupLocker.js]:
```
this.onCompleteTimeout = setTimeout(this.props.onComplete, 300);

// then, clearTimeout(this.onCompleteTimeout) on unmount
```

File: src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        } else {
+            this.setState({ passwordStrength: score, lockerSuggestion: suggestions, lockerWarning: warning, goodEnough: false });
+        }
+        this.validatePassword();
+    };
+
+    handlePasswordChange = (event) => {
+        const { disableContinue } = this.state;
+
+        if (this.passwordInputTimeout != null) {
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: Please cancel this timeout when unmounting

Comment (satazor) [reply src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: Also, there's no need for the if

File: src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+        this.password = event.target.value;
+        !disableContinue && this.setState({ disableContinue: true });
+
+        this.passwordInputTimeout = setTimeout(this.analyzePassword, 400);
+    };
+
+    handleConfirmationChange = (event) => {
+        const { disableContinue } = this.state;
+
+        if (this.confirmationInputTimeout != null) {
```
Comment (satazor) [inline src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: Same as above

File: src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js (hunk 15 above)
Comment (satazor) [reply src/shared/components/setup-locker/setup-locker-steps/set-passphrase/SetPassphrase.js]: Maybe `validatePasswordMatch`
